packages/pds/src/pipethrough.ts

```diff
@@ -367,80 +418,135 @@ function handleUpstreamRequestError(
+function isAllowedAccept(accept: Accept): boolean {
+  return accept[1].q > 0
 }
 
-function nameIncludedIn(this: readonly string[], accept: Accept): boolean {
-  return this.includes(accept[0])
+/**
+ * @see {@link https://developer.mozilla.org/en-US/docs/Glossary/Quality_values}
+ */
+function formatAcceptHeader(accept: readonly [Accept, ...Accept[]]): string {
+  return accept.map(formatAcceptPart).join(',')
 }
 
-function isNotRejected(accept: Accept): boolean {
-  return accept[1]['q'] !== '0'
+function formatAcceptPart(accept: Accept): string {
+  return `${accept[0]};q=${accept[1].q}`
 }
 
-function extractAcceptedNames(
-  acceptHeader: undefined | string | string[],
-): string[] {
-  if (!acceptHeader?.length) {
-    return ['*']
+function parseAcceptEncoding(
+  acceptEncodings: undefined | string | string[],
+): undefined | Accept[] {
+  if (acceptEncodings == null) return undefined
+
+  if (typeof acceptEncodings !== 'string') {
+    // For type-safety. We should never get here
+    throw new InvalidRequestError('accept-encoding must be a string')
   }
 
-  return Array.isArray(acceptHeader)
-    ? acceptHeader.flatMap(extractAcceptedNames)
-    : acceptHeader.split(',').map(extractAcceptedName).filter(isNonNullable)
+  return acceptEncodings.split(',').map(parseAcceptEncodingDefinition)
 }
 
-function extractAcceptedName(def: string): string | undefined {
-  // No need to fully parse since we only care about allowed values
-  const parts = def.split(';')
-  if (parts.some(isQzero)) return undefined
-  return parts[0].trim()
-}
+function parseAcceptEncodingDefinition(def: string): Accept {
+  const { length, 0: encoding, 1: params } = def.trim().split(';', 3)
 
-function isQzero(def: string): boolean {
-  return def.trim() === 'q=0'
-}
+  if (length > 2) {
+    throw new InvalidRequestError(`Invalid accept-encoding: "${def}"`)
+  }
+
+  if (!encoding || encoding.includes('=')) {
+    throw new InvalidRequestError(`Invalid accept-encoding: "${def}"`)
+  }
+
+  const flags = { q: 1 }
+  if (length === 2) {
+    const { length, 0: key, 1: value } = params.split('=', 3)
+    if (length !== 2) {
+      throw new InvalidRequestError(`Invalid accept-encoding: "${def}"`)
+    }
+
+    if (key === 'q' || key === 'Q') {
+      const q = parseFloat(value)
+      if (q === 0 || (Number.isFinite(q) && q <= 1 && q >= 0.001)) {
+        flags.q = q
+      } else {
+        throw new InvalidRequestError(`Invalid accept-encoding: "${def}"`)
+      }
+    } else {
+      throw new InvalidRequestError(`Invalid accept-encoding: "${def}"`)
+    }
+  }
 
-function isNonNullable<T>(val: T): val is NonNullable<T> {
-  return val != null
+  return [encoding.toLowerCase(), flags]
 }
 
-export function isJsonContentType(contentType?: string): boolean | undefined {
+export function isJsonContentType(
+  contentType?: string | string[],
+): boolean | undefined {
   if (contentType == null) return undefined
+  if (typeof contentType !== 'string') return undefined
```

Comment:
I don't think `isJsonContentType()` is ever called with a `string[]`— do we need this change?

Reply:
Letfover form intermediary changes I think. Removed.